contributors/design-proposals/node/troubleshoot-running-pods.md

```diff
@@ -192,96 +225,110 @@ type EphemeralContainer struct {
+runtime label. Persisting this data in the runtime means it will survive kubelet
+restarts.
+
+At least for the Docker runtime, this is [an intended use of docker
+labels](https://docs.docker.com/engine/userguide/labels-custom-metadata/#value-guidelines).
+Docker does not document the maximum length of labels in its API. Empirically,
+it supports up to the 64K constraint of the docker client's `bufio.Scanner`
+size. Because the container spec may be examined in security sensitive contexts
+like admission control, we will conservatively limit the size of the spec to 32K
+and add a 32K minimum label length test to runtime qualification.
+
+`EphemeralContainer` is populated by the kubelet in the same way as regular
+container statuses. This is sent to the API server and displayed by `kubectl
+describe pod`.
 
 ### Creating Debug Containers
 
-1.  `kubectl` invokes the exec API as described in the preceding section.
-1.  The API server checks for name collisions with existing containers, performs
+1.  `kubectl` invokes the new API as described in the preceding section.
+1.  The API server checks for name collisions with existing containers (in both
+    `PodSpec` and `PodStatus.EphemeralContainers`), performs validation,
     admission control and proxies the connection to the kubelet's
-    `/exec/$NS/$POD_NAME/$CONTAINER_NAME` endpoint.
-1.  The kubelet instructs the Runtime Manager to create a Debug Container.
-1.  The runtime manager uses the existing `startContainer()` method to create a
-    container in an existing pod. `startContainer()` has one modification for
-    Debug Containers: it creates a new runtime label (e.g. a docker label) that
-    identifies this container as a Debug Container.
-1.  After creating the container, the kubelet schedules an asynchronous update
-    of `PodStatus`. The update publishes the debug container status to the API
-    server at which point the Debug Container becomes visible via `kubectl
-    describe pod`.
-1.  The kubelet will upgrade the connection to streaming and attach to the
-    container's console.
-
-Rather than performing the implicit attach the kubelet could return success to
-the client and require the client to perform an explicit attach, but the
-implicit attach maintains consistent semantics across `/exec` rather than
-varying behavior based on parameters.
+    `/ephemeralContainers/$NS/$POD_NAME` endpoint.
+    1.  Since a name collision could happen in the interval between container
+        creation and PodStatus being published to the API server, the kubelet
+        will perform an additional check for name collision.
+1.  The kubelet request handler opens an error channel and signals the pod's
+    sync worker with `UpdatePodOptions` that include the `EphemeralContainer` in
+    a new field and a callback in the existing
+    `UpdatePodOptions.OnCompleteFunc`.
+    1.  The pod sync worker runs the existing `syncPod()` with a new
```

Comment:
The debug containers won't be sync'd, except to update ContainerStatus. After being started it will be ignored until the pod sandbox dies (or is garbage collected). Clarified.